beluga_ros/include/beluga_ros/amcl_impl.hpp

```diff
@@ -0,0 +1,215 @@
+                         return std::make_pair(result.x(), result.y());
+                       }) |
+                       ranges::to<std::vector>;
+
+    std::visit(
+        [&, this](auto& execution_policy, auto& motion_model, auto& sensor_model) {
+          update_particles(
+              execution_policy,                                           //
+              motion_model(control_action_window_ << base_pose_in_odom),  //
+              sensor_model(std::move(measurement)));
+        },
+        execution_policy_, motion_model_, sensor_model_);
+    force_update_ = false;
+    return beluga::estimate(beluga::views::states(particles_), beluga::views::weights(particles_));
+  }
+
+  /// Force a manual update of the particles on the next iteration of the filter.
+  void force_update() { force_update_ = true; }
+
+ private:
+  beluga::TupleVector<particle_type> particles_;
+
+  AmclImplParams params_;
+  beluga::MultivariateUniformDistribution<Sophus::SE2d, beluga_ros::OccupancyGrid> map_distribution_;
+  motion_model_variant motion_model_;
+  sensor_model_variant sensor_model_;
+  execution_policy_variant execution_policy_;
+
+  beluga::spatial_hash<Sophus::SE2d> spatial_hasher_;
+  beluga::ThrunRecoveryProbabilityEstimator random_probability_estimator_;
+  beluga::any_policy<Sophus::SE2d> update_policy_;
+  beluga::any_policy<decltype(particles_)> resample_policy_;
+
+  beluga::RollingWindow<Sophus::SE2d, 2> control_action_window_;
+
+  bool force_update_{true};
+
+  std::optional<std::pair<Sophus::SE2d, Eigen::Matrix3d>> latest_estimate_;
+  std::optional<Sophus::SE2d> latest_odom_transform_in_map_;
+
+  double random_state_probability_{0.0};
+
+  template <class ExecutionPolicy, class MotionUpdate, class SensorUpdate>
+  void update_particles(ExecutionPolicy&& policy, MotionUpdate motion_update, SensorUpdate sensor_update) {
+    particles_ |= beluga::actions::propagate(policy, std::move(motion_update)) |  //
+                  beluga::actions::reweight(policy, std::move(sensor_update)) |   //
+                  beluga::actions::normalize(policy);
+
+    random_state_probability_ = random_probability_estimator_(particles_);
+
+    if (resample_policy_(particles_)) {
+      auto random_state = ranges::compose(beluga::make_from_state<particle_type>, std::ref(map_distribution_));
+
+      if (random_state_probability_ > 0.0) {
+        random_probability_estimator_.reset();
+      }
+
+      particles_ |= beluga::views::sample |
+                    beluga::views::random_intersperse(std::move(random_state), random_state_probability_) |
+                    beluga::views::take_while_kld(
+                        spatial_hasher_,        //
+                        params_.min_particles,  //
+                        params_.max_particles,  //
+                        params_.kld_epsilon,    //
+                        params_.kld_z) |
+                    beluga::actions::assign;
+    }
+  }
```

Comment:
@nahueespinosa meta: does this need to be here? It is independent from the template method parameters. If not, do we need this method at all? `AMCLImpl::update()` wouldn't be that much more complex if we inline `update_particles` over there (and that's kinda the point of all these changes we did).

Reply:
Totally, fixed in https://github.com/Ekumen-OS/beluga/pull/327/commits/6ab472d30c99d5d6230bf505f6ad20fb963c0bfb.